extensions/cba_settings/cba_settings.cpp

```diff
@@ -0,0 +1,335 @@
+            result = result + trim(line) + "\n";
+    }
+
+    file.close();
+
+    if (strlen * MAX_CALLBACK_LENGTH > result.length())
+        return "";
+
+    result = result.substr(strlen * MAX_CALLBACK_LENGTH, MAX_CALLBACK_LENGTH);
+    return result;
+}
+
+string parse(int strlen) {
+    string result = "[";
+
+    ifstream file;
+    file.open(SETTINGS_FILE, ios::in);
+
+    string line;
+    while (file.good()) {
+        getline(file, line);
+        line = trim_left(line);
+
+        // Ignore commented out lines.
+        if (line.substr(0, 2) == "//")
+            continue;
+
+        // Count force tags.
+        int force = 0;
+        while (line.substr(0, 6) == "force ") {
+            force++;
+            line = line.substr(6);
+            line = trim_left(line);
+        }
+
+        // Get setting.
+        size_t i = line.find("=");
+        if (i == string::npos) // Ignore ill-formed line.
+            continue;
+
+        string item = line.substr(0, i);
+        item = trim_right(item);
+
+        // Get value.
+        string value = line.substr(i + 1);
+        value = trim(value);
+
+        while (value.back() == ';')
+            value.pop_back();
+
+        value = trim_right(value);
+
+        result = result + "[\"" + item + "\"," + value + "," + to_string(force) + "],";
+    }
+
+    file.close();
+
+    result.pop_back();
+    result.push_back(']');
+
+    if (strlen * MAX_CALLBACK_LENGTH > result.length())
+        return "";
+
+    result = result.substr(strlen * MAX_CALLBACK_LENGTH, MAX_CALLBACK_LENGTH);
+    return result;
+}
+
+string read(string setting) {
+    string result = "[\"\",false,-1]";
+
+    ifstream file;
+    file.open(SETTINGS_FILE, ios::in);
+
+    string line;
+    while (file.good()) {
+        getline(file, line);
+        line = trim_left(line);
+
+        // Ignore commented out lines.
+        if (line.substr(0, 2) == "//")
+            continue;
+
+        // Count force tags.
+        int force = 0;
+        while (line.substr(0, 6) == "force ") {
+            force++;
+            line = line.substr(6);
+            line = trim_left(line);
+        }
+
+        // Get setting.
+        size_t i = line.find("=");
+        if (i == string::npos) // Ignore ill-formed line.
+            continue;
+
+        string item = line.substr(0, i);
+        item = trim_right(item);
+
+        if (_stricmp(item.c_str(), setting.c_str()) == 0) {
+            // Get value.
+            string value = line.substr(i + 1);
+            value = trim(value);
+
+            while (value.back() == ';')
+                value.pop_back();
+
+            value = trim_right(value);
+
+            result = "[\"" + item + "\"," + value + "," + to_string(force) + "]";
+            break;
+        }
+    }
+
+    file.close();
+    return result;
+}
+
+string append(string setting, string value, int force) {
```

Comment:
Passing by reference is advisable with classes. It  avoids copying the whole class. The same can be applied to the other functions.

```suggestion
string append(const string &setting, const string &value, int force) {
```

Reply:
Cool. Didn't know it copies, and that this is how one prevents it.